site/docs/plugins/chat-members.md

```diff
@@ -0,0 +1,132 @@
+
+bot.use(chatMembers(adapter));
+
+bot.start({
+  allowed_updates: ["chat_member", "message"],
+  onStart: ({ username }) => console.log(`Listening as ${username}`),
+});
+```
+
+</CodeGroupItem>
+
+<CodeGroupItem title="JavaScript" active>
+
+```js
+import { Bot, Context, MemorySessionStorage } from "grammy";
+import { chatMembers, ChatMembersFlavor } from "@grammyjs/chat-members";
+
+const adapter = new MemorySessionStorage();
+
+const bot = new Bot("<your bot token>");
+
+bot.use(chatMembers(adapter));
+
+bot.start({
+  allowed_updates: ["chat_member", "message"],
+  onStart: ({ username }) => console.log(`Listening as ${username}`),
+});
+```
+
+</CodeGroupItem>
+
+<CodeGroupItem title="Deno" active>
+
+```ts
+import {
+  Bot,
+  Context,
+  MemorySessionStorage,
+} from "https://deno.land/x/grammy/mod.ts";
+import type { ChatMember } from "https://deno.land/x/grammy/types.ts";
+import {
+  chatMembers,
+  ChatMembersFlavor,
+} from "https://deno.land/x/grammy_chat_members/mod.ts";
+
+type MyContext = Context & ChatMembersFlavor;
+
+const adapter = new MemorySessionStorage<ChatMember>();
+
+const bot = new Bot<MyContext>("<your bot token>");
+
+bot.use(chatMembers(adapter));
+
+bot.start({
+  allowed_updates: ["chat_member", "message"],
+  onStart: ({ username }) => console.log(`Listening as ${username}`),
+});
+```
+
+</CodeGroupItem>
+
+</CodeGroup>
+
+### Reading Chat Members
+
+This plugin also adds a new `ctx.chatMembers.getChatMember` function that will check the storage for information about a chat member before querying Telegram for it.
+If the chat member exists in the storage, it will be returned.
+Otherwise, `ctx.api.getChatMember` will be called and the result will be saved to the storage, making subsequent calls faster and removing the need to call Telegram again for that user and chat in the future.
+
+Here's an example:
+
+```ts
+bot.on("message", async (ctx) => {
+  const chatMember = await ctx.chatMembers.getChatMember();
+
+  return ctx.reply(
+    `Hello, ${chatMember.user.first_name}! I see you are a ${chatMember.status} of this chat!`,
+  );
+});
+```
+
+All parameters are optional; if you don't provide them, `ctx.chat.id` and `ctx.from.id` will be used instead.
+Please notice that, if you don't provide a chat id and there's no `chat` property inside the context (for example, on inline query updates), this will throw an error.
+The same will happen if there's no `ctx.from` in the context.
+
+## Aggressive Storage
+
+The `enableAggressiveStorage` config option will install middleware to cache chat members without depending on the `chat_member` event.
+For every update, the middleware checks if `ctx.chat` and `ctx.from` exist.
+If they both do, it then proceeds to call `ctx.chatMembers.getChatMember` to add the chat member information to the storage in case it doesn't exist.
+
+Please note that this means the storage will be called for **every update**, which may be a lot, depending on how many updates your bot receives.
```

Comment:
```suggestion
Please note that this means the storage will be written to for **every update**, which may be a lot, depending on how many updates your bot receives.
```

Reply:
It won't be written if the information is already there. It is only guaranteed to be read, to check if the record exists.

Reply:
Is this better?

```suggestion
Please note that this means the storage will be read for **every update**, which may be a lot, depending on how many updates your bot receives.
```

Reply:
Maybe I am misunderstanding something here. I thought this:

Default:
- write to storage on chat_member
- read from storage as requested

Aggressive mode:
- write to storage for all updates that contain ctx.chat or ctx.from
- read from storage as requested

Is this accurate?

Reply:
Not really. Every time the plugin runs, it has to check to see if the current chat member already exists in the database, so it needs to do at least one read operation. If the member does not exist yet, then it does a write operation. By default, it only runs on `chat_member` updates. Aggressive mode makes it so it runs on every update, effectively doing a database read for every update.

Reply:
One thing we could do with the new storage adapter methods is keep an in-memory list of the known keys.
We could load all the keys on plugin setup and then always update both the in-memory list and the database whenever needed.